test/e2e/fishapp/dynamic_stack.go

```diff
@@ -0,0 +1,811 @@
+//		-> /path2 -> vn1(50)
+//				  -> vn2(50)
+//	vs4 -> /path1 -> vn3(50)
+//				  -> vn4(50)
+//		-> /path2 -> vn5(50)
+//				  -> vn6(50)
+//	vs5 -> /path1 -> vn7(50)
+//				  -> vn8(50)
+//		-> /path2 -> vn9(50)
+//				  -> vn10(50)
+// =======virtual nodes =========
+//  vn1 -> vs1,vs2,vs3,vs4
+//  vn2 -> vs5,vs1,vs2,vs3
+//  vn3 -> vs4,vs5,vs1,vs2
+//  ...
+//
+// then we validate each virtual node can access each virtual service at every path, and calculates the target distribution
+type DynamicStack struct {
+	// service discovery type
+	ServiceDiscoveryType shared.ServiceDiscoveryType
+
+	// number of virtual service
+	VirtualServicesCount int
+
+	// number of virtual nodes count
+	VirtualNodesCount int
+
+	// number of routes per virtual router
+	RoutesCountPerVirtualRouter int
+
+	// number of targets per route
+	TargetsCountPerRoute int
+
+	// number of backends per virtual node
+	BackendsCountPerVirtualNode int
+
+	// number of replicas per virtual node
+	ReplicasPerVirtualNode int32
+
+	// how many time to check connectivity per URL
+	ConnectivityCheckPerURL int
+
+	// ====== runtime variables ======
+	mesh              *appmeshv1beta1.Mesh
+	namespace         *corev1.Namespace
+	cloudMapNamespace string
+
+	createdNodeVNs  []*appmeshv1beta1.VirtualNode
+	createdNodeDPs  []*appsv1.Deployment
+	createdNodeSVCs []*corev1.Service
+
+	createdServiceVSs  []*appmeshv1beta1.VirtualService
+	createdServiceSVCs []*corev1.Service
+}
+
+// expects the stack can be deployed to namespace successfully
+func (s *DynamicStack) Deploy(ctx context.Context, f *framework.Framework) {
+	s.createMeshAndNamespace(ctx, f)
+	if s.ServiceDiscoveryType == shared.CloudMapServiceDiscovery {
+		s.createCloudMapNamespace(ctx, f)
+	}
+	mb := &shared.ManifestBuilder{
+		MeshName:             s.mesh.Name,
+		Namespace:            s.namespace.Name,
+		ServiceDiscoveryType: s.ServiceDiscoveryType,
+		CloudMapNamespace:    s.cloudMapNamespace,
+	}
+	s.createResourcesForNodes(ctx, f, mb)
+	s.createResourcesForServices(ctx, f, mb)
+	s.grantVirtualNodesBackendAccess(ctx, f)
+}
+
+// expects the stack can be cleaned up from namespace successfully
+func (s *DynamicStack) Cleanup(ctx context.Context, f *framework.Framework) {
+	var deletionErrors []error
+	if errs := s.revokeVirtualNodeBackendAccess(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	if errs := s.deleteResourcesForServices(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	if errs := s.deleteResourcesForNodes(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	if s.ServiceDiscoveryType == shared.CloudMapServiceDiscovery {
+		if errs := s.deleteCloudMapNamespace(ctx, f); len(errs) != 0 {
+			deletionErrors = append(deletionErrors, errs...)
+		}
+	}
+	if errs := s.deleteMeshAndNamespace(ctx, f); len(errs) != 0 {
+		deletionErrors = append(deletionErrors, errs...)
+	}
+	Expect(len(deletionErrors)).To(BeZero())
```

Comment:
What happens if the cleanup step fails? Do resources need to be manually removed, or can you retry cleanup only?

Reply:
the cleanup steps are independent from each other(failed steps will report errors).
And yes, if resources are left over, we need to manually removed. I don't think add retry helps before we saw resource leaks scenarios. (like apiserver crash etc).
We'll monitor test failures and use dedicated account to run tests. If we saw resource leaks, we'll add handling for them correspondingly.